spec/abbreviato/abbreviato_spec.rb

```diff
@@ -0,0 +1,330 @@
+      with: { max_length: 14 },
+      source: "𠝹𠝹",
+      expected: "<p>𠝹...</p>"
+  end
+
+  describe "multi-byte tail" do
+    # These examples purposely specify a number of bytes which is not divisible by four, to ensure
+    # characters don't get brokwn up part-way thorugh their multi-byte representation
+    it_does_not_truncate "html text with longer length",
+      with: { max_length: 99, tail: '𠴕' },
+      source: "<p>𠲖𠲖𠲖𠲖𠲖𠲖𠲖𠲖</p>",
+      expected: "<p>𠲖𠲖𠲖𠲖𠲖𠲖𠲖𠲖</p>"
+    it_does_not_truncate "html text with equal length",
+      with: { max_length: 15, tail: "𠴕" },
+      source: "<p>𠝹𠝹</p>",
+      expected: "<p>𠝹𠝹</p>"
+    it_truncates "html text with shorter length",
+      with: { max_length: 15, tail: "𠴕" },
+      source: "<p>𠝹𠝹𠝹𠝹</p>",
+      expected: "<p>𠝹𠴕</p>"
+    it_truncates "html text with shorter length and longer tail",
+      with: { max_length: 23, tail: "𠴕𠴕𠴕" },
+      source: "<p>𠝹𠝹𠝹𠝹𠝹𠝹</p>",
+      expected: "<p>𠝹𠴕𠴕𠴕</p>"
+  end
+
+  describe "html entity (ellipsis) tail" do
+    it_truncates "html text with ellipsis html entity tail",
+      with: { max_length: 27, tail: '&hellip;' },
+      source: "<p>This is some text which will be truncated</p>",
+      expected: "<p>This is some&hellip;</p>"
+  end
+
+  describe "html tags structure" do
+    it_truncates "html text with tag",
+      with: { max_length: 14 },
+      source: "<p>some text</p>",
+      expected: "<p>some...</p>"
+
+    it_truncates "html text with nested tags (first node)",
+      with: { max_length: 22 },
+      source: "<div><p>some text 1</p><p>some text 2</p></div>",
+      expected: "<div><p>s...</p></div>"
+
+    it_truncates "html text with nested tags (second node)",
+      with: { max_length: 46 },
+      source: "<div><p>some text 1</p><p>some text 2</p></div>",
+      expected: "<div><p>some text 1</p><p>some te...</p></div>"
+
+    it_truncates "html text with nested tags (empty contents)",
+      with: { max_length: 13 },
+      source: "<div><p>some text 1</p><p>some text 2</p></div>",
+      expected: "<div></div>"
+
+    it_truncates "html text with special html entities",
+      with: { max_length: 15 },
+      source: "<p>&gt;some text</p>",
+      expected: "<p>&gt;s...</p>"
+
+    it_truncates "html text with siblings tags",
+      with: { max_length: 64 },
+      source: "<div>some text 0</div><div><p>some text 1</p><p>some text 2</p></div>",
+      expected: "<div>some text 0</div><div><p>some text 1</p><p>som...</p></div>"
+
+    it_does_not_truncate "html with unclosed tags",
+      with: { max_length: 151 },
+      source: "<table><tr><td>Hi <br> there</td></tr></table>",
+      expected: "<table><tr><td>Hi <br/> there</td></tr></table>"
+
+    it_does_not_truncate "preserve html entities",
+      with: { max_length: 99 },
+      source: "<o:p>&nbsp;</o:p>",
+      expected: "<o:p>&nbsp;</o:p>"
+
+    it_truncates "not truncate html entities (all or nothing)",
```

Comment:
@rmatrz Technically doesn't contradict, but I've improved the description to make it clearer what's being tested.